midgen.go

```diff
@@ -0,0 +1,45 @@
+package pubsub
+
+import (
+	"sync"
+)
+
+// msgIDGenerator handles computing IDs for msgs
+// It allows setting custom generators(MsgIdFunction) per topic
+type msgIDGenerator struct {
+	Default MsgIdFunction
+
+	topicGens   map[string]MsgIdFunction
+	topicGensLk sync.RWMutex
```

Comment:
nit: can you put the lock on top so that it is clear that it protects the member underneat?
This pattern is called the mutex hat.

Reply:
Mutex hat, like this. Done!